src/redprl/refiner.fun

```diff
@@ -1015,8 +1015,224 @@ struct
+      else if dimsContradictory (ext0, eps0) then [] (* second row, left rule *)
+      else if dimsContradictory (ext1, eps1) then [] (* second row, left rule *)
+      else
+        let
+          val (P.VAR u, P.APP _) = (ext0, eps0)
+          val (P.VAR v, P.APP _) = (ext1, eps1)
+        in
+          if Sym.eq (u, v) andalso not (P.eq Sym.eq (eps0, eps1)) then [] (* second row, right rule *)
+          else
+            [Seq.map (substSymbol (eps0, u)) (Seq.map (substSymbol (eps1, v)) jdg)] (* bottom right rule *)
+        end
+  end
+
+  structure HCom =
+  struct
+    (* Given a list of n extents and a list of 2n tubes, produce a list of
+       (extent * side * tube) triples, where side is 0 or 1 according to the
+       tube's position in the list *)
+    fun groupTubes [] [] = []
+      | groupTubes (ext :: exts) (tube0 :: tube1 :: tubes)  =
+        (ext, P.APP P.DIM0, tube0) :: (ext, P.APP P.DIM1, tube1) :: groupTubes exts tubes
+      | groupTubes _ _ = raise Fail "groupTubes"
+
+    fun listToTel l = List.foldl (fn (g, l) => l >: g) T.empty l
+
+    (* Produce the list of goals requiring that tube aspects agree with each other. *)
+    fun intraTubeGoals H w ty group =
+      let
+        fun intraTube (ext0, eps0, (u, tube0)) (ext1, eps1, (v, tube1)) =
+          let
+            val tube0 = substSymbol (P.ret w, u) tube0
+            val tube1 = substSymbol (P.ret w, v) tube1
+            val J = H >> CJ.EQ ((tube0,tube1), ty)
+          in
+            List.map (fn j => #1 (makeGoal (([(w, P.DIM)], []) |> j)))
+                     (Restriction.Two J ext0 eps0 ext1 eps1)
+          end
+      in
+        listToTel
+          (ListMonad.bind
+             (fn x => ListMonad.bind (intraTube x) group)
+             group)
+      end
+
+    (* Produce the list of goals requiring that tube aspects agree with the cap. *)
+    fun tubeCapGoals H ty r cap group =
+      let
+        fun tubeCap (ext, eps, (u, tube)) =
+          let
+            val J = H >> CJ.EQ ((substSymbol (r,u) tube, cap), ty)
+          in
+            List.map (#1 o makeGoal) (Restriction.One J ext eps)
+          end
+      in
+        listToTel (ListMonad.bind tubeCap group)
+      end
+
+    fun Eq alpha jdg =
+      let
+        val _ = RedPrlLog.trace "HCom.Eq"
+        val H >> CJ.EQ ((lhs, rhs), ty) = jdg
+        val Syn.HCOM (exts0, (r0, r'0), ty0, cap0, tubes0) = Syn.out lhs
+        val () = assertAlphaEq (ty0, ty)
+        val Syn.HCOM (exts1, (r1, r'1), ty1, cap1, tubes1) = Syn.out rhs
+        val () = assertParamEq "HCom.Eq source of direction" (r0, r1)
+        val () = assertParamEq "HCom.Eq target of direction" (r'0, r'1)
+        val _ = ListPair.mapEq (assertParamEq "HCom.Eq extents") (exts0, exts1)
+
+        val (goalTy, _) = makeGoal @@ H >> CJ.EQ_TYPE (ty0, ty1)
+        val (goalCap, _) = makeGoal @@ H >> CJ.EQ ((cap0, cap1), ty)
+
+        val w = alpha 0
+
+        fun interTube ext eps (u, tube0) (v, tube1) =
```

Comment:
Can you add a comment here which describes what this does, analogous to the comment on `intraTubeGoals`? In these comments, it might also be nice to write (in ascii) an example premise that they correspond to from the paper.